examples/bert/run_pretraining.py

```diff
@@ -398,16 +419,33 @@ def main(_):
         num_warmup_steps=FLAGS.num_warmup_steps,
         num_train_steps=FLAGS.num_train_steps,
     )
+    optimizer = keras.optimizers.Adam(learning_rate=learning_rate_schedule)
+
+    if FLAGS.restore_from_checkpoint is not None:
```

Comment:
I think we need to make this work so it restores after a failure in a automatic fashion.

If the checkpoint save path is set, we should save a checkpoint after each epoch (is this currently saving best checkpoint or one for each epoch?).

If the script is terminated and re-run with the same arguments, we automatically pick up where we left off. And then we should maybe add a `skip_restore` flag to avoid this behavior, and always start from scratch, that default to false.

Reply:
+1, it is very important that training should be fully autonomous: there will be failures, and we do NOT want to have to rerun the script with different manually-specified arguments every time there is a failure. The restart should be automated, and should resume from the latest saved state (both model wise and data pipeline wise) and that state should be retrievable without modifying any command line argument.

Reply:
The file loading logic looks a little fragile to me. Have you took a look at these:

 - https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/BackupAndRestore
 - https://www.tensorflow.org/api_docs/python/tf/train/CheckpointManager (lower level)

We should make sure we are modeling the best practices for doing this with the example we have here. Seems like there should be a better flow than this one here?

Reply:
Changed!